src/multibase.js

```diff
@@ -0,0 +1,127 @@
+'use strict'
+
+const constants = require('./constants')
+const bs58 = require('bs58')
+
+exports = module.exports = multibase
+exports.encode = encode
+exports.decode = decode
+exports.isEncoded = isEncoded
+
+const errNotSupported = new Error('Unsupported encoding')
+
+// returns a new buffer with the multibase varint+code`
+function multibase (nameOrCode, buf) {
+  if (!buf) {
+    throw new Error('requires an encoded buffer')
+  }
+  const code = getCode(nameOrCode)
+  const codeBuf = new Buffer(code)
+
+  const name = getName(nameOrCode)
+  validEncode(name, buf)
+  return Buffer.concat([codeBuf, buf])
+}
+
+function encode (nameOrCode, buf) {
+  const name = getName(nameOrCode)
+
+  let encode
+
+  switch (name) {
+    case 'base58btc': {
+      encode = (buf) => { return new Buffer(bs58.encode(buf)) }
+    } break
+    default: throw errNotSupported
+  }
+
+  return multibase(name, encode(buf))
+}
+
+// receives a buffer or string encoded with multibase header
+// decodes it and returns an object with the decoded buffer
+// and the encoded type { base: <name>, data: <buffer> }
+function decode (bufOrString) {
+  if (Buffer.isBuffer(bufOrString)) {
+    bufOrString = bufOrString.toString()
+  }
+
+  const code = bufOrString.substring(0, 1)
+  bufOrString = bufOrString.substring(1, bufOrString.length)
+
+  if (typeof bufOrString === 'string') {
+    bufOrString = new Buffer(bufOrString)
+  }
```

Comment:
Still, this would be much cleaner without the back and forth conversions

```js
let code
let rest
if (Buffer.isBuffer(in)) {
  code = in.toString('utf8', 0, 1)
  rest = in.slice(1)
} else {
  code = in.substring(0, 1)
  rest = Buffer(in.slice(1))
}
```